### PR TITLE
feat!: update models to cat-vrs 1.0.0 trial review ballot

### LINKED ISSUE #8: Update models to 1.0.0 Trial Review Ballot (November 2024)
### Feature description

A new ballot was released here: https://github.com/ga4gh/cat-vrs/releases/tag/1.0.0-ballot.2024-11.1

We should update the Pydantic models to align with those in the spec release

### Use case

As a user, I want to be able to use the latest models

### Proposed solution

_No response_

### Alternatives considered

_No response_

### Implementation details

_No response_

### Potential Impact

_No response_

### Additional context

_No response_

### Contribution

Yes, I can create a PR for this feature.

### PR DESCRIPTION
**Pull request by korikuzma**
close #8 

* Update models to cat-vrs [1.0.0-ballot.2024-11.2](https://github.com/ga4gh/cat-vrs/tree/1.0.0-ballot.2024-11.2) tag

### CODE REVIEW

**Comment by korikuzma** [comment]
Probably good for me to add some tests for the recipe validators

**Comment by korikuzma** [comment]
@brendanreardon Would like your review too

File: src/ga4gh/cat_vrs/recipes.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        Exactly one constraint in ``constraints`` must satisfy ALL of the following
+        requirements:
+        1. Must be a ``DefiningAlleleConstraint``
+        2. Must have ``relations`` property that meets ALL of the following
+        requirements:
+            a. Must contain exactly one relation where ``primaryCode = liftover_to``
+            b. Must contain exactly one relation where ``primaryCode = transcribes_to``
+
+        :param v: Constraints property to validate
+        :raises ValueError: If constraints property does not satisfy the requirements
```
Comment (korikuzma) [inline src/ga4gh/cat_vrs/recipes.py:90]: @brendanreardon @DanielPuthawala do you like this approach or would you rather make all checks and list ALL validation errors at once?

Comment (brendanreardon) [reply src/ga4gh/cat_vrs/recipes.py:90]: This is so cool!! Thank you, @korikuzma! 

This makes sense to me, and I think that it is preferable to list the validation errors as they come to enhance tracebackability (this that a word?).

**Comment by brendanreardon** [comment]
> @brendanreardon Would like your review too

Aye, aye! Once I get back into building out examples, I'll try running the current recipes through your code here.

File: src/ga4gh/cat_vrs/recipes.py (hunk 1 above)
Comment (korikuzma) [reply src/ga4gh/cat_vrs/recipes.py:90]: Okay, we can always change in a future version if we want to list all errors at once (which is nice for development)

Comment (brendanreardon) [reply src/ga4gh/cat_vrs/recipes.py:90]: Ah gotcha. In that case since the latter is better for development, I totally defer to you!

Comment (korikuzma) [reply src/ga4gh/cat_vrs/recipes.py:90]: Okay, let's leave it as is for now (to be consistent with VRS-Python too).

**Review by DanielPuthawala — APPROVED** [review]
Looks good.  we talked through the changes earlier, I jst forgot to come back on and submit my review.

File: src/ga4gh/cat_vrs/models.py
[the remark this answers is not in the document]
Comment (korikuzma) [reply src/ga4gh/cat_vrs/models.py:1]: @DanielPuthawala I don't think I'm following... Can you rephrase?

[the remark this answers is not in the document]
Comment (DanielPuthawala) [reply src/ga4gh/cat_vrs/models.py:1]: That was a question I had before about constraints haveng an explicit enum of relations to instead having just a slot for relations and having the actual list elsewhere (in GKS:core).

You and I later on discussed this, and decided it wasn't something we had to address for this PR.

[the remark this answers is not in the document]
Comment (korikuzma) [reply src/ga4gh/cat_vrs/models.py:1]: Ah okay. Thank you

File: src/ga4gh/cat_vrs/recipes.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        :return: Constraints property
+        """
+        constraint = v[0]
```
Comment (brendanreardon) [inline src/ga4gh/cat_vrs/recipes.py:93]: Here, this presumes that the list of constraints is of length 1. Should there be a similar "if not" statement checking the length of the `list[Constraint]` input? e.g.,

```python
if not len(v) == 1:
	err_msg = "CanonicalAllele requires a minimum and maximum of 1 constraint"
	raise ValueError(err_msg)
```

Without this error check, users could pass a list of constraints for a canonical allele and this function will only check the first constraint. 

With that said, this will _probably_ be changed in a future branch per adding FeatureContextConstraint.

File: src/ga4gh/cat_vrs/recipes.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        max_length=2,
+        description="The constraints array must contain exactly two items: a DefiningLocationConstraint and either a CopyChangeConstraint or CopyCountConstraint.",
+    )
+
+    @field_validator("constraints")
+    @classmethod
+    def validate_constraints(cls, v: list[Constraint]) -> list[Constraint]:
+        """Validate constraints property
+
+        Each constraint in ``constraints`` must satisfy ALL of the following
```
Comment (brendanreardon) [inline src/ga4gh/cat_vrs/recipes.py:143]: This docstring does not fully reflect the schema for [CategoricalCnv](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/schema/cat-vrs/recipes-source.yaml#L70-L97), I think. In particular, the requirements specified here apply to `constraints` / `v`, not each constraint in `constraints`.

If so, this docstring should instead say
```
``constraints`` must contain two constraints that satisfy ALL of the following requirements, respectively:
```

This could alternatively read
```
``constraints`` must contain two constraints - ``DefiningLocationConstraint`` and either ``CopyCountConstraint`` or ``CopyChangeCount`` - that satisfy ALL of the following requirements, respectively:
```

but I think that this second one seems to be more against the style that these docstrings have been written.

File: src/ga4gh/cat_vrs/recipes.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        requirements:
+        1. Must contain a ``DefiningLocationConstraint`` where
+            a. ``relations`` property contains at least one relation where ``primaryCode = liftover_to``
+        2. Must contain a ``CopyCountConstraint`` or ``CopyChangeConstraint``
+
+        :param v: Constraints property to validate
+        :raises ValueError: If constraints property does not satisfy the requirements
+        :return: Constraints property
+        """
+        def_loc_constr_found = False
```
Comment (brendanreardon) [inline src/ga4gh/cat_vrs/recipes.py:153]: The error handling for `DefiningLocationConstraint` can be made more granular, similar to `CanonicalAllele` by doing something like the code below. At the moment, there is only one error message for DefiningLocationConstraint, regardless of if the relation requirement is satisfied. You can totally imagine some poor developer head banging at their desk thinking, "but I _am_ passing a DefiningLocationConstraint with the correct relations"... when maybe their relations are not set up properly.

```python
def_loc_constr_found = False  
def_loc_constr_valid = False
copy_constr_found = False  

for constraint in v:  
    constraint = constraint.root  
    if not def_loc_constr_found and isinstance(  
        constraint, DefiningLocationConstraint  
    ):  
	    def loc_constr_found = True
        for r in constraint.relations:  
            if (  
                r.primaryCode  
                and r.primaryCode.root == Relation.LIFTOVER_TO.value  
            ):  
                def_loc_constr_valid = True  
                continue  
    if not copy_constr_found:  
        copy_constr_found = isinstance(  
            constraint, CopyCountConstraint | CopyChangeConstraint  
        )  
  
if not def_loc_constr_found:  
    err_msg = f"Must contain a `DefiningLocationConstraint` with at least one relation where `primaryCode` is '{Relation.LIFTOVER_TO.value}'."  
    raise ValueError(err_msg)  

if not def_loc_constr_valid:
	err_msg = f"``DefiningLocationConstraint found``, but Must contain at least one relation where `primaryCode` is '{Relation.LIFTOVER_TO.value}'"
	raise ValueError(err_msg)

if not copy_constr_found:  
    err_msg = (  
        "Must contain either a `CopyCountConstraint` or `CopyChangeConstraint`."  
    )  
    raise ValueError(err_msg)
```

File: tests/validation/test_cat_vrs_models.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    valid_params = deepcopy(members)
+    valid_params["constraints"] = [
+        models.Constraint(
+            root=models.DefiningAlleleConstraint(
+                relations=[
+                    MappableConcept(
+                        primaryCode=code(models.Relation.TRANSLATES_FROM.value)
+                    )
+                ],
+                allele="allele.json#/1",
```
Comment (brendanreardon) [inline tests/validation/test_cat_vrs_models.py:73]: Does this refer to [Allele.json](https://github.com/ga4gh/vrs/blob/0106b48c1db0b420efd787cc769e4b35f415068e/schema/vrs/json/Allele) in vrs' schemas? If so, should this be capitalized?

File: tests/validation/test_cat_vrs_models.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        copy_change_constr,
+    ]
+    with pytest.raises(
+        ValueError,
+        match="Must contain a `DefiningLocationConstraint` with at least one relation where `primaryCode` is 'liftover_to'.",
+    ):
+        recipes.CategoricalCnv(**invalid_params)
+
+    # Invalid No CopyCountConstraint or CopyChangeConstraint
+    invalid_params["constraints"] = [
```
Comment (brendanreardon) [inline tests/validation/test_cat_vrs_models.py:342]: This should instead be

```python
invalid_params["constraints"] = [
    copy_change_constr
]
```

right? This may also require some additional tests to make sure that either `CopyCountConstraint` or `CopyChangeConstraint`, whichever is passed, is valid.

**Review by brendanreardon — CHANGES_REQUESTED** [review]
Thank you for this, @korikuzma !! @DanielPuthawala may need to look over these pieces of feedback

File: src/ga4gh/cat_vrs/recipes.py (hunk 2 above)
Comment (korikuzma) [reply src/ga4gh/cat_vrs/recipes.py:93]: @brendanreardon So in line 74 we have: `constraints: list[Constraint] = Field(..., min_length=1, max_length=1)` to ensure this list has a length of 1

File: tests/validation/test_cat_vrs_models.py (hunk 5 above)
Comment (korikuzma) [reply tests/validation/test_cat_vrs_models.py:73]: Oh, this was just a dummy IRI. I can make a comment

File: tests/validation/test_cat_vrs_models.py (hunk 6 above)
Comment (korikuzma) [reply tests/validation/test_cat_vrs_models.py:344]: I can update the comment to be clearer. This was testing that constraints does not contain either a `CopyCountConstraint` or a `CopyChangeConstraint`. 

The first two examples in `test_categorical_cnv` test valid input (either `CopyCountConstraint` or `CopyChangeConstraint`). I have tests for `CopyChangeConstrint` in `test_copy_change_constraint`, but can make tests for `CopyCountConstraint`.

File: src/ga4gh/cat_vrs/recipes.py (hunk 3 above)
Comment (korikuzma) [reply src/ga4gh/cat_vrs/recipes.py]: @brendanreardon Updated. Let me know what you think

File: src/ga4gh/cat_vrs/recipes.py (hunk 4 above)
Comment (korikuzma) [reply src/ga4gh/cat_vrs/recipes.py:152]: Added!

File: src/ga4gh/cat_vrs/recipes.py (hunk 3 above)
Comment (brendanreardon) [reply src/ga4gh/cat_vrs/recipes.py]: I love it! Thank you!

File: tests/validation/test_cat_vrs_models.py (hunk 6 above)
Comment (brendanreardon) [reply tests/validation/test_cat_vrs_models.py:357]: re: comment Ah!! Okay, I see now. Thank you! 

re: tests. that was so quick! thank you!

File: tests/validation/test_cat_vrs_models.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+    return models.DefiningLocationConstraint(
+        relations=[
+            MappableConcept(primaryCode=code(models.Relation.LIFTOVER_TO.value))
+        ],
+        location="location.json#/1",
+        matchCharacteristic=MappableConcept(label="test"),
+    )
+
+
+def test_copy_count_constraint():
```
Comment (brendanreardon) [inline tests/validation/test_cat_vrs_models.py:53]: this looks great!

**Review by brendanreardon — APPROVED** [review]
🤜 🤛 this looks great! Sorry again for punting on this

**Comment by DanielPuthawala** [comment]
Just caught up on your guys' back-and-forth.  It still looks good to me.  Are you going to squash and merge now, Kori?
